### PR TITLE
feat: add get/set all columns, fixes #546

### LINKED ISSUE #546: BUG: hiding a column by which the table is sorted breaks `multi column sorting`
please use the following example:
1. remove the `.txt` suffix, 
1. put it in the `examples` folder
1. order by `% complete`, `start`. 
1. hide `start` column.
1. add `finish` to sort

voila! check your browser's console, and you will see the message! the behavior is visualized in the following GIF.

**the reason** behind this behavior is that `slick-grid` changes the entire column-set each time a column is set to be hidden; which causes the `getColumnIndex` function to return `null`, as the hidden column is not on the list of columns; obviously! 

**the solution** is more complicated as far as I can tell! because many plugins and existing codes depend on the current implementation of the `getColumns` and `getColumnIndex` functions, it's not easy to add a property - for example, called `visible` ` to the column definition in order to control column visibility. hiding or showing columns shouldn't be a matter or setting a new column-set, but rather a matter of changing their visibility and redrawing the whole table!

multi-sort-issue

[example-multi-column-sort.html-broken.txt](https://github.com/6pac/SlickGrid/files/5444730/example-multi-column-sort.html-broken.txt)

### PR DESCRIPTION
**Pull request by ghiscoding**
- supersede #549

As per previous PR comments:
> added new property `allColumns` and 3 new functions `setAllColumns(columnDefinitions)` `getAllColumns(defaultToColumns)` and getColumnById(columnId, useVisibleColumns)`- the last function is used in`setupColumnSort` in order to get correct column: if `allColumns` is set then use `allColumns` else use `columns` as reference list

### CODE REVIEW

**Comment by ghiscoding** [comment]
@6pac 
This is the last PR I wanted to do before a new release. Could you please merge this and then proceed with a `minor` release `3.1.0`

**Comment by 6pac** [comment]
Could I have just  quick summary of how you use this? For example, is there a ```.visible``` property to determine which rows are displayed? I can't see one. If not, could you give a quick example of how hidden rows work.

**Comment by ghiscoding** [comment]
the method `grid.getColumns()` doesn't always return all columns defined, if for example the user calls `grid.setColumn([...])` with less columns, this in terms will hide columns that are part of the array provided to `setColumns`. For example in my libs, I have to keep all columns in a variable somewhere so that I can for example change locale and I need to translate all column titles even the hidden columns. Hidden are typically used with the Column Picker and/or Grid Menu, you can hide and eventually show them back, if we hide some columns and we use `getColumsn()` then we would only get the visible columns but what about the hidden ones? These new methods helps with that, to get all columns and that is including the hidden ones

in summary
- `getColumns()` returns only the visible columns
- `getAllColumns()` returns all visible + all hidden

**Comment by 6pac** [comment]
OK, and so I'm assuming that ```setColumns``` sets all columns if no columns have been set so far, and just the visible columns if ```setAllColumns``` has been used on a larger array of columns.

I'm sorry to make comments at this late stage, but I do find this approach over-complex. What if someone sets 'All Columns' and then passes in a 'Columns' list that includes a column that is not in the 'AllColumns' list?

Why not a simple ```.visible``` property on the column objects, defaulting to ```true```. This would be way cleaner and simpler. I'm happy to submit that if you like, I've been thinking of adding it for a while.

**Comment by ghiscoding** [comment]
For more info on how it works, it basically saves all column definitions into that new variable when we initialize the grid, because we never initialize the grid with hidden columns, we always have them at that time and so we can assume that is the best time to save that info and if the user calls `setColumns()` after then we know it might not include all columns

**Comment by 6pac** [comment]
I note that at the moment, I also keep an external 'all' column list where I need to hide or show columns.

**Comment by ghiscoding** [comment]
> I'm sorry to make comments at this late stage, but I do find this approach over-complex. What if someone sets 'All Columns' and then passes in a 'Columns' list that includes a column that is not in the 'AllColumns' list?
> 
> Why not a simple `.visible` property on the column objects, defaulting to `true`. This would be way cleaner and simpler. I'm happy to submit that if you like, I've been thinking of adding it for a while.

hmm that would require to change the current code implementation of `getColumns()` and `setColumns()`, while the new methods were simply added without changing the current methods. If you want to do that instead, then I guess that yes it would probably be less lines of code, the way we did it was to add new methods to make sure we didn't break the old ones but if you are confident to do the change then go ahead, we have enough Cypress tests to cover the change, just make sure to create a PR and not push directly to master if you want to tackle it :P

it's bedtime for me now, passed 1am, cheers

**Comment by ghiscoding** [comment]
@6pac 
Hey so just to make sure that we're not introducing a breaking change with the `visible` property that you talked about. I would still expect the `getColumns()` and `setColumns()` to only affect the hidden columns. I think that to not be a breaking change, we could perhaps have `getColumns(includeHidden?: boolean)` where the `includeHidden` is optional, so this way it wouldn't break anyone's code (including mine) and something similar could be done on the setter `setColumns(cols, includeHidden?: boolean)` because I would still expect that if I call it the old way then it would still hide some columns in the grid

**Comment by 6pac** [comment]
OK, let's thrash this out, then :-)

As far as I can see, if I add a ```.visible``` property to the column that defaults to ```true```, then because all current code doesn't know about that flag, all columns will always be visible, which is non-breaking. The only way to have non-visible columns is to have an external array, which is what we all do now.

If someone wants to use the flag to make columns invisible, then they use it aware of what it will do. 

```getColumns()``` and ```setColumns``` always get and set all the columns, like they do now. If a column is hidden, then it will be in the column list, but it will have ```.visible===false```

**Comment by ghiscoding** [comment]
> EDIT: ah, I see what you mean now - if someone uses the feature, existing plugins might not work because they assume that `getColumns()` is always the visible columns. TBH, I think we need to rewrite all of the plugins in that case. I can handle the ones in this repo - it should be fairly simple to search for `getColumns()` and `setColumns()`. How much work iwould it be for your downstream repos?

Hmm this PR isn't something that I really need, it was mainly to close that old PR and these methods were a nice addition. If you want to do the rewrite then I would suggest to probably wait to do it after the jQuery removal PR is merged (it needs a lot of cleanup though). I wouldn't mind to close this PR and revisit it in the future with another PR of yours... if so could you please do the release with what we have already merged, I'd really like to be done with my libs too.

**Comment by 6pac** [comment]
No worries. I'll prepare this and another feature I've been thinking of to make the variable height rows more flexible. 

As separate PRs, to be applied after things stabilise.

**Comment by ghiscoding** [comment]
I guess that the more changes we do in `slick.grid.js` and `slick.dataview.js`, the more conflicts there will be in that jQuery removal PR. It should be fine if the changes are small but something to consider when it's a lot of changes
